src/core/thread/mle.cpp

```diff
@@ -2829,11 +2841,24 @@ void Mle::HandleAdvertisement(const Message &aMessage, const Ip6::MessageInfo &a
 void Mle::HandleDataResponse(const Message &aMessage, const Ip6::MessageInfo &aMessageInfo, const Neighbor *aNeighbor)
 {
     otError error;
+#if OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE
+    Tlv      tlv;
+    uint16_t metricsReportOffset;
+#endif
 
     Log(kMessageReceive, kTypeDataResponse, aMessageInfo.GetPeerAddr());
 
     VerifyOrExit(aNeighbor && aNeighbor->IsStateValid(), error = OT_ERROR_SECURITY);
 
+#if OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE
+    if (Tlv::FindTlvOffset(aMessage, Tlv::kLinkMetricsReport, metricsReportOffset) == OT_ERROR_NONE)
+    {
+        aMessage.Read(metricsReportOffset, sizeof(tlv), &tlv);
+        Get<LinkMetrics>().HandleLinkMetricsReport(aMessage, metricsReportOffset + sizeof(tlv), tlv.GetLength(),
```

Comment:
We can use `Tlv::FindTlvValueOffset()` here to directly get the offet of value and its length to pass to `HandleLinkMetricsReport()` (no need to read the `tlv`).